### ISSUE TITLE
[Bug]: Out of Memory (OOM) Issues During MMLU Evaluation with lm_eval

### ISSUE THREAD

**Issue by wchen61** [vllm-project/vllm#10325]
### Your current environment

vllm 0.6.0
lm_eval 0.4.5
torch 2.4
A100 + CUDA 12.3

### 🐛 Describe the bug

**Description:**
When using lm_eval for MMLU accuracy evaluation tasks, I frequently encounter OOM errors. This issue seems to be model-specific, and many models are prone to this problem. For example, even when running the Meta-Llama-8B-Instruct model on an A100 GPU (https://huggingface.co/meta-llama/Meta-Llama-3-8B-Instruct), OOM errors still occur.

I have verified that offline inference with this model works fine, meaning both my software and hardware are capable of running the model. However, when using lm_eval, the system encounters OOM issues. Specifically, when the batch size is set to auto, the system behaves similarly to the benchmark_throughput scenario: all requests are placed in the pool, and vllm continuously fetches requests for inference, followed by result analysis.

Upon further investigation, I discovered that the discrepancy between the maximum memory usage reported by the [profile_run ](https://github.com/vllm-project/vllm/blob/32e7db25365415841ebc7c4215851743fbb1bad1/vllm/worker/model_runner.py#L1045C9-L1045C20) function and the actual maximum memory usage is due to the different sampling parameters used in lm_eval. Specifically, lm_eval uses the [prompt_logprobs=1](https://github.com/EleutherAI/lm-evaluation-harness/blob/0845b588303f1f59af98dd1c5bdbd78a9e75a1e2/lm_eval/models/vllm_causallms.py#L232) setting, which causes a significant increase in memory consumption. For example, with `max-num-seqs=256` and `max-num-batched-tokens=8096`, the default configuration reports a peak memory usage of 10GB, but with` prompt_logprobs=1`, the peak memory reaches 50GB. Our system reserves memory based on the `profile_run `peak, which leads to OOM errors during actual execution.

The command line I used to run lm_eval is as follows.
>lm_eval --model vllm --model_args pretrained=meta-llama/Meta-Llama-3-8B-Instruct,tensor_parallel_size=1,dtype=auto,gpu_memory_utilization=0.9 --tasks mmlu  --batch_size auto

When I manually set `prompt_logprobs=1` in the vllm sampling parameters, lm_eval runs successfully.

**Suggested Improvement:**

It would be helpful to introduce a mechanism that allows third-party users to specify their use case, so vllm can more accurately estimate the required peak memory usage.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by hemildesai** [comment]
I faced a similar issue. Making the change suggested above in [profile_run](https://github.com/vllm-project/vllm/blob/32e7db25365415841ebc7c4215851743fbb1bad1/vllm/worker/model_runner.py#L1045C9-L1045C20) fixed it for me as well.

**Comment by zhaocaibei123** [comment]
I met the same problem. I wonder why `prompt_logprobs=1` takes so much memory, only output with logit(one float) may not consume 40GB?

**Comment by fxmarty-amd** [comment]
From what I have seen, `prompt_logprobs=1` results in the output logits not being filtered in https://github.com/vllm-project/vllm/blob/c81081fece240736e30e0f9a5ed82bb5b483c561/vllm/model_executor/layers/logits_processor.py#L64 and thus sampling occurs on the full sequence instead of just generated tokens, I guess, which appears to be memory intensive.

Adding `prompt_logprobs=1` to https://github.com/vllm-project/vllm/blob/c81081fece240736e30e0f9a5ed82bb5b483c561/vllm/worker/model_runner.py#L1250 is not enough for me to solve the issue, in my case I would OOM during warmup as the default `max_model_length` was 131072, which OOMs in the lm head. I had to agressively reduce `--max-model-len`.

There might be room to reduce the memory pressure in sampling.
